### PR TITLE
Mert Miftar 5238

### PR DESCRIPTION
**Pull request by mertm27**

### CODE REVIEW

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 1]
```diff
@@ -0,0 +1,254 @@
+// Define interfaces for movie data
+interface Movie {
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:2]: Correct interfaces +10

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        genre: string[];
+        plot: string;
+        cast: { 
+            actor: string; character: string 
+        }[];
+        oscars: { [key: string]: string };
+        rating: number;
+    }
+    
+    type MovieSorter = (first: Movie, second: Movie) => number;
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:16]: Function type for the sorter functions, +5

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    
+    document.addEventListener("DOMContentLoaded", siteCode);
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:18]: Using DOMContentLoaded +5

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        castItem.textContent = `${actor.actor} as ${actor.character}`;
+        castList.appendChild(castItem);
+        });
+        
+        modalContent.appendChild(castList);
+        modal.classList.remove("hidden");
+    }
+    
+    async function loadData() {
+        const dataUri =
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:244]: Loading data properly +10

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        genres.forEach(genre => {
+        const option = document.createElement("option");
+        option.value = genre;
+        option.textContent = genre;
+        genreFilter.appendChild(option);
+        });
+    }
+    
+    function displayMovies(movies: Movie[]) {
+        const container = document.getElementById("movie-container")!;
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:155]: Two level management of movie display +5

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        
+        const genreFilter = document.getElementById("genre-filter") as
+        HTMLSelectElement;
+        genreFilter.addEventListener("change", applyFilter);
+        
+        const oscarsFilter = document.getElementById("oscar-filter") as
+        HTMLSelectElement;
+        oscarsFilter.addEventListener("change", applyFilter);
+        
+        const idSort = document.getElementById("sort-id")!;
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:44]: Attaching events handlers for sort elements, +5

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    
+    async function siteCode() {
+        const data = await loadData();
+        movies = data;
+        
+        fillGenreDropdown(movies);
+        
+        displayMovies(movies);
+        
+        const titleFilter = document.getElementById("title-filter") as HTMLInputElement;
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:30]: Attaching events handlers for filter element +5

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+        const genreSort = document.getElementById("sort-genre")!;
+        genreSort.addEventListener("click", sortByGenre);
+        
+        const oscarsSort = document.getElementById("sort-oscars")!;
+        oscarsSort.addEventListener("click", sortByOscars);
+        
+        const ratingSort = document.getElementById("sort-rating")!;
+        ratingSort.addEventListener("click", sortByRating);
+        
+        const modal = document.getElementById("cast-details")!;
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:65]: Modal for cast details extra +5

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        modal.addEventListener("click", () => {
+        modal.classList.add("hidden");
+        });
+    }
+    
+    const titleSorter: MovieSorter = (first, second) =>
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:71]: Using sorter functions with event handlers +10

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+        const oscarsMatch = oscarsFilter === "all" || movie.oscars?.[oscarsFilter];
+        
+        return titleMatch && yearMatch && genreMatch && oscarsMatch;
+        });
+        
+        displayMovies(filteredMovies);
+    }
+    
+    function fillGenreDropdown(movies: Movie[]) {
+        const genreFilter = document.getElementById("genre-filter") as
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:136]: Perfect way to fill the genre dropdown +10

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        displayMovies(sortedMovies);
+    }
+    
+    function sortByRating() {
+        const sortedMovies = movies.slice().sort((a, b) => a.rating - b.rating);
+        displayMovies(sortedMovies);
+    }
+    
+    function applyFilter() {
+        const titleFilter = (document.getElementById("title-filter") as
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:114]: Getting filter values +10

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+        HTMLInputElement).value.toLowerCase();
+        const yearFilter = parseInt((document.getElementById("year-filter") as
+        HTMLInputElement).value, 10);
+        const genreFilter = (document.getElementById("genre-filter") as
+        HTMLSelectElement).value;
+        const oscarsFilter = (document.getElementById("oscar-filter") as
+        HTMLSelectElement).value;
+        
+        const filteredMovies = movies.filter(movie => {
+        const titleMatch = movie.title.toLowerCase().includes(titleFilter);
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:124]: Actual filtering +15

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+        const container = document.getElementById("movie-container")!;
+        container.innerHTML = "";
+        for (const movie of movies) {
+        const movieRow = generateMovieRow(movie);
+        container.appendChild(movieRow);
+        }
+    }
+    
+    function generateMovieRow(movie: Movie) {
+        const row = document.createElement("div");
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:164]: Generating elements and displaying movie data +15

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+        row.appendChild(plotCell);
+        row.appendChild(castCell);
+        row.appendChild(oscarsCell);
+        row.appendChild(ratingCell);
+        
+        return row;
+    }
+    
+    function createMovieCell(value: string) {
+        const cell = document.createElement("div");
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:191]: Nice function extraction Extra +5

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+        cell.classList.add("movie-data");
+        cell.textContent = value;
+        return cell;
+    }
+    
+    function truncateText(text: string, maxLength: number) {
+        if (text.length <= maxLength) {
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:198]: Extraction of plot value +10

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+            return truncatedText.substr(0, lastSpaceIndex) + "...";
+        } else {
+            return truncatedText + "...";
+        }
+    }
+    
+    function formatCastList(cast: { actor: string; character: string }[]) {
+        const sortedCast = cast.slice().sort((a, b) => a.actor.localeCompare(b.actor));
+        
+        if (sortedCast.length > 5) {
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:215]: Nice trick :)

File: midterm-2023-11-07/code/5238-mert-miftar/script.ts [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+        
+        if (lastSpaceIndex !== -1) {
+            return truncatedText.substr(0, lastSpaceIndex) + "...";
+        } else {
+            return truncatedText + "...";
+        }
+    }
+    
+    function formatCastList(cast: { actor: string; character: string }[]) {
+        const sortedCast = cast.slice().sort((a, b) => a.actor.localeCompare(b.actor));
```
Comment (sweko) [inline midterm-2023-11-07/code/5238-mert-miftar/script.ts:213]: Extraction of cast string, +10

**Review by sweko — APPROVED** [review]
@mertm27 - 135/100

Some great solutions, but I have to wonder how you actually tested this, as I don't see any actual changes to the HTML? Was there an error during submittal?

**Comment by mertm27** [comment]
Dear professor,

Sorry for the inconvenience. While I was taking the exam I accidentally
deleted the .ts file while testing. Then I was trying to get the file back
by pressing ctrl+z and probably at that moment I undo some changes in the
HTML file. I was hurrying because it was the last minutes of the exam so I
just committed my code that I had and hoped for the best. Thank you for
your understanding!

Best regards,

**Pull request merged**